### ISSUE TITLE
Open generics in null coalescing operator

### ISSUE THREAD

**Issue by KathleenDollard** [dotnet/vblang#339]
We are considering a change to the behavior and spec for operands to null coalescing operators. The relevant part of the spec is [here](https://github.com/dotnet/vblang/blob/master/spec/expressions.md#conditional-expressions) and the effect would be allowing unconstrained types. Where we currently say:
 
> If two operands are provided, both of the operands must be classified as values, and the first operand must be either a reference type or a nullable value type.
 
We would update the spec to say:
 
> If two operands are provided, both of the operands must be classified as values, and the first operand must not be a non-nullable value type.

C# has this proposed [here](https://github.com/dotnet/csharplang/blob/master/proposals/null-coalescing-assignment.md#detailed-design) and I'd like us to tag along. 

Can anyone see a backwards compatibility issue? This seems just goodness to me.

**Comment by franzalex** [comment]
Looks good to me. With the implementation of the null coalescing operator, this is already the default behaviour so updating the docs to reflect this isn't an issue.

**Comment by 333fred** [comment]
> With the implementation of the null coalescing operator, this is already the default behaviour so updating the docs to reflect this isn't an issue.
@franzalex this is not correct. VB's implementation of the null coalescing operator correctly follows the spec, and does not allow `If(t1, t2)`, where the type of `t1` is the unconstrained type parameter `T`. Example from latest roslyn master: https://sharplab.io/#v2:DYLgbgRgNAJiDUAfAkgWwA4HsBOAXAzgAQDKAnvrgKaoCwAUAAoCuEwAlgMaEDCwAhviLd6hUYWatOJFoQCyACgDyAM0IAVAJTzcARkIBBImqiFcAJgNGNIsbYAibVIQAehALyFky7TpPnrdLYAogB2MNIQ9KHhvAL4QA===

**Comment by franzalex** [comment]
@333fred
Whoops! I having switched to C# for almost all of my projects except when the client explicitly requests for it to be done in VB.NET, I thought (quite wrongly) that it functions the same way in both languages!

**Comment by reduckted** [comment]
> Can anyone see a backwards compatibility issue?

Are you able to provide an example or two of what this would allow us to write compared to what we can do today?

> C# has this proposed [here](https://github.com/dotnet/csharplang/blob/master/proposals/null-coalescing-assignment.md#detailed-design) and I'd like us to tag along.

That proposal doesn't mention anything about open generics. Am I not interpreting it correctly, or am I missing something? 😕

**Comment by 333fred** [comment]
@reduckted it's the second part of the proposal. See my example previously for what this enables.

**Comment by KathleenDollard** [comment]
An example of something this would enable in VB (thanks @AlekseyTs)

Public Sub Foo(Of T)(x As T, y as T)
   Dim z As T = If(x, y)
End Sub

**Comment by KathleenDollard** [comment]
Note from discussion in the VB LDM

```vb
Public Sub Foo(Of T)(x As T)
  Dim z As T = If(Nothing, x)
End Sub
```

- z is always x

```c#
void Foo<T>(T x)
{
  T z = default(T) ??  x; 
}
```

- z is 0 when T is int
- z is x when T is reference type

**Comment by paul1956** [comment]
@KathleenDollard I am somewhat confused by this, according to Roslyn Source dotnet/roslyn#29724 is in VB 16 but I can't find any specification update or documentation on this change or #29724 and this is still open.
